bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java

```diff
@@ -0,0 +1,254 @@
+        configuration.ifPresent(cfg -> {
+            refreshInterval = cfg.refreshInterval;
+            logger.debug("Start scheduled task to read device in every {} seconds", refreshInterval);
+            scheduledTask = scheduler.scheduleWithFixedDelay(this::executePeridioc, CHECK_PERIOD_SEC, CHECK_PERIOD_SEC,
+                    TimeUnit.SECONDS);
+        });
+        sinceLastReadSec.set(refreshInterval); // update immediately
+    }
+
+    @Override
+    public void dispose() {
+        logger.debug("Dispose");
+        cancelScheduledTask();
+        serviceState = ServiceState.NOT_RESOLVED;
+        readState = ReadState.IDLE;
+        super.dispose();
+    }
+
+    private void cancelScheduledTask() {
+        if (scheduledTask != null) {
+            scheduledTask.cancel(true);
+            scheduledTask = null;
+        }
+    }
+
+    private void executePeridioc() {
+        sinceLastReadSec.addAndGet(CHECK_PERIOD_SEC);
+        execute();
+    }
+
+    private synchronized void execute() {
+        ConnectionState connectionState = device.getConnectionState();
+        logger.debug("Device {} state is {}, serviceState {}, readState {}", address, connectionState, serviceState,
+                readState);
+
+        switch (connectionState) {
+            case DISCOVERED:
+            case DISCONNECTED:
+                if (isTimeToRead()) {
+                    connect();
+                }
+                break;
+            case CONNECTED:
+                read();
+                break;
+            default:
+                break;
+        }
+    }
+
+    private void connect() {
+        logger.debug("Connect to device {}...", address);
+        if (!device.connect()) {
+            logger.debug("Connecting to device {} failed", address);
+        }
+    }
+
+    private void disconnect() {
+        logger.debug("Disconnect from device {}...", address);
+        if (!device.disconnect()) {
+            logger.debug("Disconnect from device {} failed", address);
+        }
+    }
+
+    private void read() {
+        switch (serviceState) {
+            case NOT_RESOLVED:
+                discoverServices();
+                break;
+            case RESOLVED:
+                switch (readState) {
+                    case IDLE:
+                        logger.debug("Read data from device {}...", address);
+                        BluetoothCharacteristic characteristic = device.getCharacteristic(uuid);
+                        if (device.readCharacteristic(characteristic)) {
+                            readState = ReadState.READING;
+                        } else {
+                            logger.debug("Read data from device {} failed", address);
+                            disconnect();
+                        }
+                        break;
+                    default:
+                        break;
+                }
+            default:
+                break;
+        }
+    }
```

Comment:
> You've made a rather complex state machine and I think it should be simplified.

I wouldn't say that this is complex.

Reply:
Maybe complex is the wrong word. Let's just say the code can't be understood at a glance.